🎉 Zstd 1.5.0 Release 🎉

senhuang42 opened the pull request:
## Changelog

* api: Various functions promoted from experimental to stable API: (#2579-#2581, @senhuang42)
  * `ZSTD_defaultCLevel()`
  * `ZSTD_getDictID_fromCDict()`
* api: Several experimental functions have been deprecated and will emit a compiler warning (#2582, @senhuang42)
  *   `ZSTD_compress_advanced()`
  * `ZSTD_compress_usingCDict_advanced()`
  * `ZSTD_compressBegin_advanced()`
  * `ZSTD_compressBegin_usingCDict_advanced()`
  * `ZSTD_initCStream_srcSize()`
  * `ZSTD_initCStream_usingDict()`
  * `ZSTD_initCStream_usingCDict()`
  * `ZSTD_initCStream_advanced()`
  * `ZSTD_initCStream_usingCDict_advanced()`
  * `ZSTD_resetCStream()`
* api: ZSTDMT_NBWORKERS_MAX reduced to 64 for 32-bit environments (#2643, @Cyan4973)
* perf: Significant speed improvements for middle compression levels (#2494, @senhuang42 @terrelln)
* perf: Block splitter to improve compression ratio, enabled by default for high compression levels (#2447, @senhuang42)
* perf: Decompression loop refactor, speed improvements on `clang` and for `--long` modes (#2614 #2630, @Cyan4973)
* perf: Reduced stack usage during compression and decompression entropy stage (#2522 #2524, @terrelln)
* bug: Make the number of physical CPU cores detection more robust (#2517, @PaulBone)
* bug: Improve setting permissions of created files (#2525, @felixhandte)
* bug: Fix large dictionary non-determinism (#2607, @terrelln)
* bug: Fix various dedicated dictionary search bugs (#2540 #2586, @senhuang42 @felixhandte)
* bug: Fix non-determinism test failures on Linux i686 (#2606, @terrelln)
* bug: Fix UBSAN error in decompression (#2625, @terrelln)
* bug: Fix superblock compression divide by zero bug (#2592, @senhuang42)
* bug: Ensure `ZSTD_estimateCCtxSize*() `monotonically increases with compression level (#2538, @senhuang42)
* doc: Improve `zdict.h` dictionary training API documentation (#2622, @terrelln)
* doc: Note that public `ZSTD_free*()` functions accept NULL pointers (#2521, @animalize)
* doc: Add style guide docs for open source contributors (#2626, @Cyan4973)
* tests: Better regression test coverage for different dictionary modes (#2559, @senhuang42)
* tests: Better test coverage of index reduction (#2603, @terrelln)
* tests: OSS-Fuzz coverage for seekable format (#2617, @senhuang42)
* tests: Test coverage for ZSTD threadpool API (#2604, @senhuang42)
* build: Dynamic library built multithreaded by default (#2584, @senhuang42)
* build: Move  `zstd_errors.h`  and  `zdict.h`  to  `lib/`  root (#2597, @terrelln)
* build: Single file library build script moved to `build/` directory (#2618, @felixhandte)
* build: Allow `ZSTDMT_JOBSIZE_MIN` to be configured at compile-time, reduce default to 512KB (#2611, @Cyan4973)
* build: Fixed Meson build (#2548, @SupervisedThinking @kloczek)
* build: `ZBUFF_*()` is no longer built by default (#2583, @senhuang42)
* build: Fix excessive compiler warnings with clang-cl and CMake (#2600, @nickhutchinson)
* build: Detect presence of `md5` on Darwin (#2609, @felixhandte)
* build: Avoid SIGBUS on armv6 (#2633, @bmwiedmann)
* cli: `--progress` flag added to always display progress bar (#2595, @senhuang42)
* cli: Allow reading from block devices with `--force` (#2613, @felixhandte)
* cli: Fix CLI filesize display bug (#2550, @Cyan4973)
* cli: Fix windows CLI `--filelist` end-of-line bug (#2620, @Cyan4973)
* contrib: Various fixes for linux kernel patch (#2539, @terrelln)
* contrib: Seekable format - Decompression hanging edge case fix (#2516, @senhuang42)
* contrib: Seekable format - New seek table-only API  (#2113 #2518, @mdittmer @Cyan4973)
* contrib: Seekable format - Fix seek table descriptor check when loading (#2534, @foxeng)
* contrib: Seekable format - Decompression fix for large offsets, (#2594, @azat)
* misc: Automatically published release tarballs available on Github (#2535, @felixhandte)

mdittmer committed:
[contrib] Support seek table-only API

Memory constrained use cases that manage multiple archives benefit from
retaining multiple archive seek tables without retaining a ZSTD_seekable
instance for each.

* New opaque type for seek table: ZSTD_seekTable.
* ZSTD_seekable_copySeekTable() supports copying seek table out of a
  ZSTD_seekable.
* ZSTD_seekTable_[eachSeekTableOp]() defines seek table API that mirrors
  existing seek table operations.
* Existing ZSTD_seekable_[eachSeekTableOp]() retained; they delegate to
  ZSTD_seekTable the variant.

These changes allow the above-mentioned use cases to initialize a
ZSTD_seekable, extract its ZSTD_seekTable, then throw the ZSTD_seekable
away to save memory. Standard ZSTD operations can then be used to
decompress frames based on seek table offsets.

The copy and delegate patterns are intended to minimize impact on
existing code and clients. Using copy instead of move for the infrequent
operation extracting a seek table ensures that the extraction does not
render the ZSTD_seekable useless. Delegating to *new* seek
table-oriented APIs ensures that this is not a breaking change for
existing clients while supporting all meaningful operations that depend
only on seek table data.

Commit:
Move block header write into compressBlock_internal()

Commit:
Add a nbSeq argument to compressSequences()
Refactor ZSTD_compressBlock_internal() to do the block header write within and add nbSeq argument to compressSequences()

terrelln committed:
Add and integrate lazy row hash strategy

Commit:
Centralize entropy statistics calculations to zstd_compress.c

terrelln committed:
[contrib][linux] Expose zstd headers to avoid duplication

Expose the zstd headers in `include/linux` to avoid struct duplication.
This makes the member names not follow Kernel style guidelines, and
exposes the zstd symbols. But, the LMKL reviewers are okay with that.

Commit:
Refactor block compression logic into single function

Commit:
Add block splitter to experimental params

Commit:
Update function names for consistency

Commit:
Add recursive block split algorithm

terrelln committed:
Add HUF_writeCTable_wksp() function

This saves ~700 bytes of stack space in HUF_writeCTable.

Commit:
Add unit tests and fuzzer param

Commit:
Refactor buildSequencesStatistics() to avoid pointer increment for superblocks

Commit:
Change 1.3.4 bugfix to be cross-compatible with superblocks and normal compression

mlindsay committed:
ZSTD_seekable_decompress() example that hangs.

Commit:
Fix various fuzzer failures: repcode history, superblocks

PaulBone committed:
If cpuinfo parsing fails fallback to sysconf

PaulBone committed:
Only set numPhysicalCores if ratio is valid

Cyan4973 committed:
Merge pull request #2113 from mdittmer/expose-seek-table

[contrib] Support seek table-only API

Commit:
Fix seekable decompress hanging

Cyan4973 committed:
fix potential leak on exit

Cyan4973 committed:
fixed const guarantees

read-only objects are properly const-ified in parameters

Cyan4973 committed:
strengthen compilation flags

Cyan4973 committed:
various minor style fixes

Commit:
Merge pull request #2516 from senhuang42/seekable_hang_fix

Seekable hang fix

Cyan4973 committed:
Merge branch 'dev' into seekTable

Cyan4973 committed:
fixed gcc conversion warnings

Cyan4973 committed:
fixed wrong assert condition

Cyan4973 committed:
fixed gcc-7 conversion warning

Cyan4973 committed:
added test case for seekTable API

and simple roundtrip test

Cyan4973 committed:
Merge branch 'seekTable' of github.com:facebook/zstd into seekTable

wjssz committed:
doc: ZSTD_free*() functions accept NULL pointer

Cyan4973 committed:
added code comments for new API ZSTD_seekTable

terrelln committed:
Reduce stack usage of ZSTD_buildCTable()

It is a stack high-point for some compression strategies and has an easy
fix. This moves the normalized count into the entropy workspace.

Cyan4973 committed:
Merge pull request #2518 from facebook/seekTable

New direct seekTable access methods

terrelln committed:
Merge pull request #2522 from terrelln/stack-reduction

Reduce stack usage of ZSTD_buildCTable()

terrelln committed:
[contrib][linux-kernel] Fix unaligned.h

Fix the `unaligned.h` shim in the tests that was causing corruption in
the tests. Note that this is a problem with the test shim, not the
kernel code.

terrelln committed:
Merge pull request #2523 from terrelln/huf-stack-reduction

Add HUF_writeCTable_wksp() function

terrelln committed:
[huf] Reduce stack usage of HUF_readDTableX2 by ~460 bytes

* Use `HUF_readStats_wksp()`
* Use workspace in `HUF_fillDTableX2*()`
* Clean up workspace usage to use a workspace struct

terrelln committed:
[fse] Reduce stack usage of FSE_decompress_wksp() by 512 bytes

* Move `counting` into the workspace
* Inrease `HUF_DECOMPRESS_WORKSPACE_SIZE` by 512 bytes

felixhandte committed:
Add Tests Checking File Permissions of Created Files

Cyan4973 committed:
Merge pull request #2521 from animalize/doc_free

doc: ZSTD_free*() functions accept NULL pointer

Commit:
Merge pull request #2517 from PaulBone/num_cores

Make the number of physical CPU cores detection more robust

felixhandte committed:
Create Files with Desired Permissions; Avoid chmod(); Remove UTIL_chmod()

felixhandte committed:
Don't Block Removing File on Being Able to Read It

`open()`'s mode bits are only applied to files that are created by the call.
If the output file already exists, but is not readable, the `fopen()` would
fail, preventing us from removing it, which would mean that the file would
not end up with the correct permission bits.

It's not clear to me why the `fopen()` is there at all. `UTIL_isRegularFile()`
should be sufficient, AFAICT.

felixhandte committed:
Attempt to Fix `stat` Format for BSDs

felixhandte committed:
Fix Build for Windows

felixhandte committed:
Add GitHub Action to Automatically Publish Release Tarballs

This commit introduces a GitHub action that is triggered on release creation,
which creates the release tarball, compresses it, hashes it, signs it, and
attaches all of those files to the release.

foxeng committed:
Fix seek table descriptor check when loading

foxeng committed:
Fix seekable test to provide valid descriptor

felixhandte committed:
Remove CircleCI Artifact Generation

terrelln committed:
[contrib][linux-kernel] Update test include stubs

Update the test include stubs so they are able to run the current zstd
version in the kernel, so I can compare stack usage.

foxeng committed:
Run generic-dev:gcc-8-asan-ubsan-testzstd on latest Ubuntu

felixhandte committed:
Maintain Artifact Name Backwards Compatibility

When the tag is `v1.2.3`, name the artifacts `zstd-1.2.3.tar*` rather than
`zstd-v1.2.3.tar*`. When the tag doesn't match, use the full tag.

felixhandte committed:
Allow a Passphrase on the Key

felixhandte committed:
Merge pull request #2535 from felixhandte/gha-release-artifacts

Add GitHub Action to Automatically Publish Release Tarballs

terrelln committed:
[contrib][freestanding] Remove tracing support

Remove tracing support from `freestanding.py` to keep things simple.

terrelln committed:
[lib][tracing] Add ZSTD_NO_TRACE macro

When defined, it disables tracing, and avoids including the header.

terrelln committed:
[lib] Bump zstd version number

Commit:
Add mem monotonicity test over srcSize

Commit:
Merge pull request #2534 from foxeng/fix-seek-descriptor-check

Fix seek table descriptor check when loading

Commit:
Clean up various functions, add debuglogging for estimate vs. actual sizes

Commit:
Fix ZSTD_dedicatedDictSearch_isSupported() requirements

ihsinme committed:
simple fix for using bit operator.

good day.
It seems to me that the developer intended to use a logical operator.
so I suggest a simple fix.

terrelln committed:
[huf][fse] Clean up workspaces

* Move `counting` to a struct in `FSE_decompress_wksp_body()`
* Fix error code in `FSE_decompress_wksp_body()`
* Rename a variable in `HUF_ReadDTableX2_Workspace`

Commit:
Merge pull request #2540 from senhuang42/fix_dds_supported

Fix dedicated dict search isSupported() requirements.

Cyan4973 committed:
Refactor prefetching for the decoding loop

Following #2545,
I noticed that one field in `seq_t` is optional,
and only used in combination with prefetching.
(This may have contributed to static analyzer failure to detect correct initialization).

I then wondered if it would be possible to rewrite the code
so that this optional part is handled directly by the prefetching code
rather than delegated as an option into `ZSTD_decodeSequence()`.

This resulted into this refactoring exercise
where the prefetching responsibility is better isolated into its own function
and `ZSTD_decodeSequence()` is streamlined to contain strictly Sequence decoding operations.
Incidently, due to better code locality,
it reduces the need to send information around,
leading to simplified interface, and smaller state structures.

SupervisedThinking committed:
meson: fix build by adding missing files

fixes https://github.com/facebook/zstd/issues/2519

Cyan4973 committed:
Merge pull request #2548 from SupervisedThinking/build_fix

meson: fix build by adding missing files

terrelln committed:
[FSE] Clean up workspace using dynamically sized struct

Commit:
Make ZSTD_estimateCCtxSize_internal() loop through all srcSize parameter sets as well

Cyan4973 committed:
fix #2549

Commit:
Enable block splitter by default when applicable

Cyan4973 committed:
Merge pull request #2550 from facebook/vvv_srcSize

fix #2549

terrelln committed:
Merge pull request #2524 from terrelln/huf-stack-reduction

[huf] Reduce stack usage of HUF_readDTableX2 by ~972 bytes

Commit:
Merge pull request #2538 from senhuang42/monotonicity_test

Add memory monotonicity test over srcSize

Commit:
Add a fallback in case the total blocksize of split blocks exceeds raw block size

Commit:
Rename *compressSequences*() functions for clarity

Cyan4973 committed:
fix man page typo

fix #2553 detected by @jwilk

Cyan4973 committed:
Merge pull request #2554 from facebook/fix2553

fix man page typo

Commit:
Merge pull request #2447 from senhuang42/block_splitter_v2

Recursive block splitting

terrelln committed:
Merge pull request #2539 from terrelln/linux-kernel-fixes

Fixes for the next linux kernel patch version

Commit:
Add ZSTD_defaultCLevel() function to public API

Commit:
Add tests to regression tests for dict

Commit:
Move lastCountSize into a returned struct, fix MSAN error

Commit:
Add clang msan fuzz test to github actions

Commit:
Merge pull request #2555 from senhuang42/default_clevel_func

Add ZSTD_defaultCLevel() function to public API

Commit:
Restrict dictmode regression tests only to advanced API, fix some compiler warnings

Commit:
Update results.csv

Commit:
Merge pull request #2558 from senhuang42/msan_block_splitter_fix

Fix block splitter minor MSAN warning.

Commit:
Merge pull request #2559 from senhuang42/add_dict_regression_tests_backup

Add different dict modes to compression ratio regression test, update results.csv

Commit:
Add enum for representing long length ID

Commit:
Merge pull request #2561 from senhuang42/longlength_enum

Add enum for representing long length ID

niclasr committed:
Stop complaining about hash tool not found

If build_dir is set the zstd build complains about md5sum not being found.
Fix this by checking if build_dir is set before checking and using the hash tool
just like in lib/Makefile .

terrelln committed:
[copyright][license] Switch to yearless copyright and some cleanup in the linux-kernel files

* Switch to yearless copyright per FB policy
* Fix up SPDX-License-Identifier lines in `contrib/linux-kernel` sources
* Add zstd copyright/license header to the `contrib/linux-kernel` sources
* Update the `tests/test-license.py` to check for yearless copyright
* Improvements to `tests/test-license.py`
* Check `contrib/linux-kernel` in `tests/test-license.py`

terrelln committed:
[contrib][linux-kernel] Add zstd_min_clevel() and zstd_max_clevel()

Commit:
Fix repcode-related OSS-fuzz issues in block splitter (#2560)

* Do not emit last partitions of blocks as RLE/uncompressed

* Fix repcode updates within block splitter

* Add a entropytables confirm function, redo ZSTD_confirmRepcodesAndEntropyTables() for better function signature

* Add a repcode updater to block splitter, no longer need to force emit compressed blocks

Commit:
[cwksp] Align all allocated "tables" and "aligneds" to 64 bytes (#2546)

* Perform 64-byte alignment of wksp tables and aligneds internally

* Clean up cwskp_finalize() function to only do two allocs

* Refactor aligned/buffer reservation code, remove ASAN req for alignment reservations

* Change from allocating 128 bytes always to allocating only buffer space as needed for tables/aligned

* Back out aligned/table reservation order restriction

* Add stricter bounds for new/resized wksps, fix comment in zstd_cwksp.h

Commit:
Update results.csv, add Row hash to regression test

terrelln committed:
Merge pull request #2541 from ihsinme/patch-1

simple fix for using bit operator.

felixhandte committed:
`rm -f` in playTests.sh

felixhandte committed:
Also Pass Mode Bits in on Windows

I think in some unix emulation environments on Windows, (cygwin?) mode bits
are somehow respected. So we might as well pass them in. Can't hurt.

Commit:
Fix dictionary force reloading clevel selection (#2570)

* Move cdict clevel override to before localdict init

* Update results.csv after dict load changes

Commit:
Maintain two repcode histories for block splitting, replace invalid repcodes (#2569)

Commit:
Move clevel override to after initLocalDict() (#2571)

Commit:
Merge pull request #2494 from senhuang42/row_hash2

SIMD Row Based Matchfinder 🚀

Commit:
Adjust nb elements to prefetch in ZSTD_row_fillHashCache()

Commit:
Correct the detection of mismatched repcodes

Commit:
Merge pull request #2572 from senhuang42/row_hash_hashcache_one_off_error_fix

Adjust nb elements to prefetch in ZSTD_row_fillHashCache()

Commit:
Merge pull request #2574 from senhuang42/repcode_mismatch_detector_fix

Correct the block splitter mismatched repcodes detection.

Commit:
Moved ZSTD_getDictID_fromCDict() into stable API

Commit:
Moved ZSTD_defaultCLevel() into stable API

Commit:
Add literalCompressionMode to stable cParams

Commit:
Remove building of ZBUFF/deprecated folder by default

Commit:
Remove ZBUFF tests

Commit:
Merge pull request #2583 from senhuang42/remove_zbuff

[1.5.0] Remove ZBUFF

Commit:
Add DDS to oss fuzzer

Commit:
Apply DDS adjustments to avoid assert failures

Commit:
Merge pull request #2586 from senhuang42/dds_fuzz

Add DDS to oss fuzzer

Commit:
Merge pull request #2579 from senhuang42/getcdictID_to_stable

[1.5.0] Promote ZSTD_getDictID_fromCDict() into stable API

Commit:
Update results.csv

terrelln committed:
[trace] Remove default definitions of weak symbols

Instead of providing a default no-op implementation, check the symbols
for `NULL` before accessing them. Providing a default implementation
doesn't reliably work with dynamic linking. Depending on link order the
default implementations may not be overridden. By skipping the default
implementation, all link order issues are resolved. If the symbols
aren't provided the weak function will be `NULL`.

Commit:
Merge pull request #2588 from senhuang42/update_regressiontest

[regressiontest] Update results.csv

terrelln committed:
Merge pull request #2589 from terrelln/tracing

[trace] Remove default definitions of weak symbols

terrelln committed:
[linux-kernel] Replace kernel-style comments

Replace kernel-style comments with regular comments.

E.g.

```
/** Before */

/* After */

/**
 * Before
 */

/*
 * After
 */

/***********************************
 * Before
 ***********************************/

/* *********************************
 * After
 ***********************************/
```

terrelln committed:
Merge pull request #2593 from terrelln/linux-comments

[linux-kernel] Replace kernel-style comments

terrelln committed:
[1.5.0] Move `zstd_errors.h` and `zdict.h` to `lib/` root

`zstd_errors.h` and `zdict.h` are public headers, so they deserve to be
in the root `lib/` directory with `zstd.h`, not mixed in with our private
headers.

Commit:
Fix chaintable check to include rowhash in ZSTD_reduceIndex()

nickhutchinson committed:
Add clang-cl build jobs to appveyor.yml

nickhutchinson committed:
CMake: fix excessive build warnings when building with clang-cl

Commit:
Merge pull request #2580 from senhuang42/defaultclevel_to_stable

[1.5.0] Promote ZSTD_defaultCLevel() into stable API

Commit:
Merge pull request #2581 from senhuang42/lcm_stable

[1.5.0] Promote ZSTD_c_literalCompressionMode to stable params

terrelln committed:
[LDM] Speed optimization on repetitive data

LDM does especially poorly on repetitive data when that data's hash happens
to have `(hash & stopMask) == 0`. Either because the `stopMask == 0` or
random chance. Optimize this case by skipping over repetitive patterns.
The detection is very simplistic, but should catch most of the offending
cases.

```
head -c 1G /dev/zero | perf stat -- ./zstd -1 -o /dev/null -v --zstd=ldmHashRateLog=1 --long
      21.187881087 seconds time elapsed

head -c 1G /dev/zero | perf stat -- ./zstd -1 -o /dev/null -v --zstd=ldmHashRateLog=1 --long
       1.149707921 seconds time elapsed

```

Commit:
Merge pull request #2598 from senhuang42/reduce_index_rowhash_fix

Fix chaintable check to include rowhash in ZSTD_reduceIndex()

terrelln committed:
Bug fix & run overflow correction much more frequently in tests

* Fix overflow correction when `windowLog < cycleLog`. Previously, we
  got the correction wrong in this case, and our chain tables and binary
  trees would be corrupted. Now, we work as long as `maxDist` is a power
  of two, by adding `MAX(maxDist, cycleSize)` to our indices.
* When `ZSTD_WINDOW_OVERFLOW_CORRECT_FREQUENTLY` is defined to non-zero
  run overflow correction as frequently as allowed without impacting
  compression ratio.
* Enable `ZSTD_WINDOW_OVERFLOW_CORRECT_FREQUENTLY` in `fuzzer` and
  `zstreamtest` as well as all the OSS-Fuzz fuzzers. This has a 5-10%
  speed penalty at most, which seems reasonable.

terrelln committed:
[tests] Reduce memory usage of MT CLI tests

Switch from `-T0` to the default `-T1` which significantly reduces
memory usage for level 19 when there are many cores. This fixes
32-bit issues of running out of address space.

Fixes #2603.

terrelln committed:
Merge pull request #2603 from terrelln/reduce-indices-fuzzer

Bug fix & run overflow correction much more frequently in tests

terrelln committed:
Merge pull request #2606 from terrelln/test-memory

[tests] Reduce memory usage of MT CLI tests

terrelln committed:
[lib] Always load the dictionary in one go

Dictionaries larger than `ZSTD_CHUNKSIZE_MAX` used to have to be loaded
in multiple segments. Instead, when we detect large dictionaries, ensure
that we reset the context's indicies. Then, for dictionaries larger than
`ZSTD_CURRENT_MAX - 1`, only load the suffix of the dictionary. Finally,
enable DDS for large dictionaries, since we no longer load in multiple
segments.

This simplifes the dictionary loading code, and reduces opportunities
for non-determinism to slip in.

Cyan4973 committed:
Documented minimum version numbers

Any stable API entry point introduced after v1.0
should be documented with its minimum version number.

Since PR fixes this requirement
updating mostly new entry points since v1.4.0
and newly introduced ones for future v1.5.0.

felixhandte committed:
Detect Presence of `md5` on Darwin

This fixes #2568.

Commit:
Fix incorrect usages of repIndex across all strategies

Cyan4973 committed:
allow jobSize to be as low as 512 KB

previous lower limit was 1 MB.

Note : by default, the lowest job size is 2 MB, achieved at level 1.
Even lower job sizes can be achieved by manipulating this value directly,
or manually modifying window sizes to lower amounts.

Updated unit test to ensure that this new limit works fine
(test would fail with previous 1 MB limit).

felixhandte committed:
Merge pull request #2609 from felixhandte/md5sum-darwin

Detect Presence of `md5` on Darwin

terrelln committed:
Merge pull request #2597 from terrelln/public-headers

[1.5.0] Move `zstd_errors.h` and `zdict.h` to `lib/` root

terrelln committed:
[easy] Rewrite rowHashLog computation

`ZSTD_highbit32(1u << x) == x` when it isn't undefined behavior.

Cyan4973 committed:
Merge pull request #2608 from facebook/docMinVer

Documented minimum version numbers

felixhandte committed:
Allow Reading from Block Devices with `--force`

felixhandte committed:
Add Test Case (Behind Flag); Run in GitHub Action

terrelln committed:
Merge pull request #2612 from terrelln/minor-fix

[easy] Rewrite rowHashLog computation

Cyan4973 committed:
fix rsyncable mode

terrelln committed:
[test] Add large dict/data --patch-from test

Dictionary size must be > `ZSTD_CHUNKSIZE_MAX`.

terrelln committed:
Merge pull request #2607 from terrelln/deterministic-dict

[lib] Always load the dictionary in one go

terrelln committed:
Merge pull request #2610 from senhuang42/lazy_underflow_fix

Fix bad integer wraparound in repcode index for fast, dfast, lazy

terrelln committed:
Merge pull request #2602 from terrelln/ldm-opt

[LDM] Speed optimization on repetitive data

Cyan4973 committed:
Merge pull request #2611 from facebook/smallerJobs

allow jobSize to be as low as 512 KB

azat committed:
seekable decompression fixes (#2594)

* seekable_format: fix from-file reading (not in-memory)

It tries to check the buffer boundary, but there is no buffer for
from-file reading.

* seekable_decompression: break when ZSTD_seekable_decompress() returns zero

* seekable_decompression_mem: break when ZSTD_seekable_decompress() returns zero

* seekable_format: cap the offset+len up to the last dOffset

This will allow to read the whole file w/o gotting corruption error if
the offset is more then the data left in file, i.e.:

    $ ./seekable_compression seekable_compression.c 8192 | head
    $ zstd -cdq seekable_compression.c.zst | wc -c
    4737

Before this patch:

    $ ./seekable_decompression seekable_compression.c.zst 0 10000000 | wc -c
    ZSTD_seekable_decompress() error : Corrupted block detected
    0

After:

    $ ./seekable_decompression seekable_compression.c.zst 0 10000000 | wc -c
    4737

Cyan4973 committed:
updated documentation regarding minimum job size

Cyan4973 committed:
Merge branch 'dev' into d_prefetch_refactor

Cyan4973 committed:
deeper prefetching pipeline for decompressSequencesLong

pipeline increased from 4 to 8 slots.
This change substantially improves decompression speed when there are long distance offsets.
example with enwik9 compressed at level 22 :
gcc-9 : 947 -> 1039 MB/s
clang-10: 884 -> 946 MB/s

I also checked the "cold dictionary" scenario,
and found a smaller benefit, around ~2%
(measurements are more noisy for this scenario).

felixhandte committed:
Merge pull request #2613 from felixhandte/allow-block-device

Allow Reading from Block Devices with `--force`

felixhandte committed:
Attempt to Fix Windows Build Error

terrelln committed:
[lib] Add ZSTD_c_deterministicRefPrefix

This flag forces zstd to always load the prefix in ext-dict mode, even
if it happens to be contiguous, to force determinism. It also applies to
dictionaries that are re-processed.

A determinism test case is also added, which fails without
`ZSTD_c_deterministicRefPrefix` and passes with it set.

Question: Should this be the default behavior? It isn't in this PR.

terrelln committed:
[lib] Move some ZSTD_CCtx_params off the stack

* Take `params` by const reference in `ZSTD_resetCCtx_internal()`.
* Add `simpleApiParams` to the CCtx and use them in the simple API
  functions, instead of creating those parameters on the stack.

I think this is a good direction to move in, because we shouldn't need
to worry about adding parameters to `ZSTD_CCtx_params`, since it should
always be on the heap (unless they become absoultely gigantic).

Some `ZSTD_CCtx_params` are still on the stack in the CDict functions,
but I've left them for now, because it was a little more complex, and we
don't use those functions in stack-constrained currently.

felixhandte committed:
Move Single-File Build Script from `contrib/` to `build/`

terrelln committed:
[tests] Set `DEBUGLEVEL=2` by default

This allows us to quickly check for compile errors in debug log
messages, which are compiled out when `DEBUGLEVEL < 2`.

felixhandte committed:
Merge pull request #2525 from felixhandte/fix-file-permissions-again

Improve Setting Permissions of Created Files

felixhandte committed:
Rewrite References to Location

Cyan4973 committed:
improved benchmark experience on Windows

benchmark results are not progressively displayed on Windows terminal.
For long benchmark sessions, nothing is displayed,
until the end, where everything is flushed.

Force display to be flushed after each update.
Updates happen roughtly every second, or even less,
so it's not a substantial workload.

Cyan4973 committed:
Merge pull request #2614 from facebook/dlong8

faster speed for decompressSequencesLong

Cyan4973 committed:
fix --filelist compatibility with Windows cr+lf line ending

terrelln committed:
[test][regression] Update results.csv

The LDM change in PR #2602 changed the algorithm slightly.
The compressed size is generally positive, and when it is worse,
it is only a few bytes.

terrelln committed:
Merge pull request #2615 from terrelln/stack-space

[lib] Move some ZSTD_CCtx_params off the stack

terrelln committed:
[zdict] Add a FAQ to the top of zdict.h

The FAQ covers the questions asked in Issue #2566. It first covers why
you would want to use a dictionary, then what a dictionary is, and
finally it tells you how to train a dictionary, and clarifies some of
the parameters.

There is definitely more that could be said about some of the advanced
trainers, but this should be a good start.

Cyan4973 committed:
Merge pull request #2619 from facebook/winbench

improved benchmark experience on Windows

Cyan4973 committed:
attempt to make Appveyor's Cygwin test faster

Cygwin is the longest Appveyor test
Appveyor is typically the CI which finish last

Commit:
Add seekable roundtrip fuzzer (#2617)

terrelln committed:
Merge pull request #2621 from terrelln/regression-test

[test][regression] Update results.csv

terrelln committed:
Merge pull request #2616 from terrelln/deterministic-dict

[lib] Add ZSTD_c_deterministicRefPrefix

felixhandte committed:
Merge pull request #2618 from felixhandte/single-file-build-mv

Move Single-File Build Script from `contrib/` to `build/`

Cyan4973 committed:
Merge pull request #2620 from facebook/winFilelist

fix --filelist compatibility with Windows cr+lf line ending

Commit:
Add --progress flag (#2595)

terrelln committed:
Merge pull request #2622 from terrelln/zdict-api

[zdict] Add a FAQ to the top of zdict.h

terrelln committed:
Merge pull request #2600 from nickhutchinson/clang-cl

Fix for excessive compiler warnings when building with clang-cl

Commit:
[1.5.0] Deprecate some functions (#2582)

* Add deprecated macro to zstd.h, mark certain functions as deprecated

* Remove ZSTD_compress.c dependencies on deprecated functions

terrelln committed:
[lib] Fix UBSAN warning in ZSTD_decompressSequences()

Cyan4973 committed:
Merge branch 'dev' into fasterCygwin

Cyan4973 committed:
switch to clang

Cyan4973 committed:
added a paragraph on coding style

terrelln committed:
Merge pull request #2625 from terrelln/ubsan-failure

[lib] Fix UBSAN warning in ZSTD_decompressSequences()

Cyan4973 committed:
Merge pull request #2626 from facebook/codingStyle1

added a paragraph on coding style

Cyan4973 committed:
Merge branch 'dev' into d_prefetch_refactor

Cyan4973 committed:
do not install g++

terrelln committed:
[lib] Fix fuzzer timeouts by backing off overflow correction

Linearly back off the frequency of overflow correction based on the
number of times the `ZSTD_window_t` has been overflow corrected. This
will still allow the fuzzer to quickly find overflow correction bugs,
while also keeping good speed for larger inputs.

Additionally, the `nbOverflowCorrections` variable can be useful for
debugging coredumps, since we can inspect the `ZSTD_CCtx` to see if
overflow correction has happened yet.

I've verified this fixes the timeouts in OSS-Fuzz (176 seconds -> 6
seconds). I've also verified that fuzzers and `fuzzer` and `zstreamtest`
still catch the row-hash overflow correction bug.

skitt committed:
Apply flags to libzstd-nomt in libzstd style

... for consistency (this doesn't actually change the build flags used
in practice, currently).

Signed-off-by: Stephen Kitt <steve@sk2.org>

Commit:
[1.5.0] Enable multithreading in lib build by default (#2584)

* Update lib Makefile to have new targets

* Update lib/README.md for mt

Commit:
Add threadPool unit tests to fuzzer.c (#2604)

Cyan4973 committed:
Merge pull request #2623 from facebook/fasterCygwin

attempt to make Appveyor's Cygwin test faster

Cyan4973 committed:
Merge branch 'dev' into d_prefetch_refactor

Commit:
Assert no divison by 0, correct superblocks 0 sequences case (#2592)

terrelln committed:
Merge pull request #2627 from terrelln/timeout-fix

[lib] Fix fuzzer timeouts by backing off overflow correction

Commit:
Add PHONY targets to makefiles (#2629)

Cyan4973 committed:
improve decompression speed of long variant by ~+5%

changed strategy,
now unconditionally prefetch the first 2 cache lines,
instead of cache lines corresponding to the first and last bytes of the match.

This better corresponds to cpu expectation,
which should auto-prefetch following cachelines on detecting the sequential nature of the read.

This is globally positive, by +5%,
though exact gains depend on compiler (from -2% to +15%).
The only negative counter-example is gcc-9.

Cyan4973 committed:
Merge branch 'd_prefetch_refactor' of github.com:facebook/zstd into d_prefetch_refactor

Cyan4973 committed:
update decoder hot loop alignment

This seems to bring an additional ~+1.2% decompression speed
on average across 10 compilers x 6 scenarios.

Cyan4973 committed:
Merge pull request #2547 from facebook/d_prefetch_refactor

Refactor prefetching for the decoding loop

Cyan4973 committed:
improved gcc-9 and gcc-10 decoding speed

the new alignment setting is better for gcc-9 and gcc-10
by about ~+5%.

Unfortunately, it's worse for essentially all other compilers.

Make the new alignment setting conditional to gcc-9+.

Cyan4973 committed:
Merge pull request #2628 from skitt/libzstd-nomt-flags

Apply flags to libzstd-nomt in libzstd style

Cyan4973 committed:
Merge pull request #2630 from facebook/gcc9

improved gcc-9 and gcc-10 decoding speed

bmwiedemann committed:
Avoid SIGBUS on armv6

When running armv6 userspace on armv8 hardware with a 64 bit Linux kernel,
the mode 2 caused SIGBUS (unaligned memory access).
Running all our arm builds in the build farm
only on armv8 simplifies administration a lot.

Depending on compiler and environment, this change might slow down
memory accesses (did not benchmark it). The original analysis is 6 years old.

Fixes #2632

Cyan4973 committed:
Merge pull request #2633 from bmwiedemann/issue2632

Avoid SIGBUS on armv6

Commit:
Bump version to 1.5.0, rebuild documentation (#2634)

Cyan4973 committed:
updated generated man pages for v1.5.0 (#2635)

Cyan4973 reviewed (approved):
As expected, 
extended fuzzer tests started during the week-end have not found anything so far.
This seems good to go.

ghost commented:
On Windows 10, maybe this release has a performance regression.

Just replace the `lib` folder, pyzstd module unit-tests use: 3.0 sec -> 3.4 sec. (intel haswell)

ghost commented:
This change is missing from changelog:

[1.5.0] Enable multithreading in lib build by default (#2584)

Commit:
Add mt lib build to CL, shuffle around bugs section (#2638)

Commit:
Remove const data members in threadpooltest payload (#2639)

Commit:
Remove const data members in threadpooltest payload (#2639) (#2640)

Cyan4973 committed:
reduce Max nb Workers to 64 in 32-bit mode

and restored limit to 256 when in 64-bit mode
(it was reduced to 200 to give more room for 32-bit).

This should fix test instability issues
using lot of threads in 32-bit environments.

occivink committed:
fileio: clamp value of windowLog in patch-mode (#2637)

With small enough input files, the inferred value of fileWindowLog could
be smaller than ZSTD_WINDOWLOG_MIN.

This can be reproduced like so:
$ echo abc > small
$ echo abcdef > small2
$ zstd --patch-from small small2 -o patch
previously, this would fail with the error "zstd: error 11 : Parameter is out of bound"

Cyan4973 committed:
Merge branch 'dev' of github.com:facebook/zstd into dev

Commit:
Update CHANGELOG to include patch from fix (#2642)

Cyan4973 committed:
Merge pull request #2643 from facebook/workers32

reduce ZSTDMT_NBWORKERS_MAX in 32-bit mode

Cyan4973 committed:
updated meson test

hopefully, bionic will have a more recent version of python
required to install meson.

Cyan4973 committed:
Merge pull request #2644 from facebook/mesonFix

Fixed meson test on travisCI

Commit:
Add source level deprecation warning disabling to certain tests/utils (#2645)

Commit:
Remove deprecate flag for vcx (#2647)
